File: breadth_first_search.rb
require_relative 'graph'
require_relative 'queue'

def bfs
  graph = create_graph

  q = MyQueue.new
  q.push(graph)

  while q.present?
    node = q.pop
    node.print

    node.adjacents.each do |adj|
      q.push(adj)
    end
  end
end

def create_graph
  g = GraphNode.new("h")
  h = g.add_adjacent("e").add_adjacent("l")
  h.add_adjacent("p")
  h.add_adjacent("l").add_adjacent("o")
  g.add_adjacent("i")
  g.add_adjacent("a").add_adjacent("t")

  g
end

bfs

File: breadthFirstSearch.ts
/**
 * Category: GRAPHS
 * Tags: algorithms search queues
 *
 * Breadth First Search of a graph. Keeps track of visited nodes.
 */
import Graph, { GraphNode } from "./utils/graph";
import Queue from "./utils/queue";
import * as assert from "assert";

class BSTSearcher {
  graph: Graph;

  constructor(g: Graph) {
    this.graph = g;
  }

  includes(value: string): boolean {
    const queue = new Queue<GraphNode>();

    queue.enqueue(this.graph.root);

    while (!queue.isEmpty()) {
      const node = queue.dequeue();
      node.visit();
      if (node.value === value) return true;

      node.adjacents.forEach((adjacent) => {
        if (adjacent.visited) return;

        queue.enqueue(adjacent);
      });
    }

    return false;
  }
}

const searcher = new BSTSearcher(Graph.generateForDijkstra());

assert(searcher.includes("Tokyo") === true);
assert(searcher.includes("Nepal") === true);
assert(searcher.includes("Seattle") === false);
console.log("All assertions passed.");
